crates/l2/operator/proof_data_provider.rs

```diff
@@ -37,49 +40,111 @@ impl ProofDataProvider {
         Self { ip, port }
     }
 
-    pub fn start(&self) {
+    pub async fn start(&self) {
         let listener = TcpListener::bind(format!("{}:{}", self.ip, self.port)).unwrap();
 
-        let mut current_id = 1;
+        let mut last_proved_block = 0;
 
         info!("Starting TCP server at {}:{}", self.ip, self.port);
         for stream in listener.incoming() {
             let stream = stream.unwrap();
 
             debug!("Connection established!");
-            self.handle_connection(stream, &mut current_id);
+            self.handle_connection(stream, &mut last_proved_block).await;
         }
     }
 
-    fn handle_connection(&self, mut stream: TcpStream, current_id: &mut u32) {
+    async fn handle_connection(&self, mut stream: TcpStream, last_proved_block: &mut u64) {
         let buf_reader = BufReader::new(&stream);
 
-        let data: ProofData = serde_json::de::from_reader(buf_reader).unwrap();
+        let data: Result<ProofData, _> = serde_json::de::from_reader(buf_reader);
         match data {
-            ProofData::Request {} => {
-                self.handle_request(&mut stream, *current_id);
-                *current_id = (*current_id % 20) + 1;
+            Ok(ProofData::Request {}) => {
+                if let Err(e) = self.handle_request(&mut stream, *last_proved_block).await {
+                    warn!("Failed to handle request: {e}");
+                }
+            }
+            Ok(ProofData::Submit { id, proof }) => {
+                if let Err(e) = self.handle_submit(&mut stream, id, proof) {
+                    warn!("Failed to handle submit: {e}");
+                }
+                *last_proved_block += 1;
+            }
+            Err(e) => {
+                warn!("Failed to parse request: {e}");
+            }
+            _ => {
+                warn!("Invalid request");
             }
-            ProofData::Submit { id, proof } => self.handle_submit(&mut stream, id, proof),
-            _ => {}
         }
 
         debug!("Connection closed");
     }
 
-    fn handle_request(&self, stream: &mut TcpStream, current_id: u32) {
+    async fn get_last_block_number() -> Result<u64, String> {
+        let response = Client::new()
+            .post("http://localhost:8551")
+            .header("content-type", "application/json")
+            .body(
+                r#"{
+                    "jsonrpc": "2.0",
+                    "method": "eth_blockNumber",
+                    "params": [],
+                    "id": 1
+                }"#,
+            )
+            .send()
+            .await
+            .map_err(|e| e.to_string())?
```

Comment:
Nit: Here you can map only the received json, and on the let statement below  match with Ok(Rpc...) or Err(e...)